uefi/src/fs/path/mod.rs

```diff
@@ -0,0 +1,53 @@
+//! This module offers the [`Path`] and [`PathBuf`] abstractions.
+//!
+//! # Interoperability with Rust strings
+//!
+//! For the interoperability with Rust strings, i.e., `String` and `str` from
+//! the standard library, the API is intended to transform these types first to
+//! `CString16` respectively `CStr16`. They do not directly translate to
+//! [`Path`] and [`PathBuf`].
+//!
+//! # Path Structure
+//!
+//! Paths use the [`SEPARATOR`] character as separator. Paths are absolute and
+//! do not contain `.` or `..` components. However, this can be implemented in
+//! the future.
+
+mod path;
+mod pathbuf;
+mod validation;
+
+pub use path::Path;
+pub use pathbuf::PathBuf;
+
+use uefi::{CStr16, Char16};
+pub(super) use validation::validate_path;
+pub use validation::PathError;
+
+/// The default separator for paths.
+pub const SEPARATOR: char = '\\';
+/// [`SEPARATOR`] but as useful UEFI type.
+pub const SEPARATOR_C16: Char16 = unsafe { Char16::from_u16_unchecked('\\' as u16) };
+
+/// Stringified version of [`SEPARATOR`].
+pub const SEPARATOR_STR: &str = "\\";
+/// [`SEPARATOR_STR`] but as useful UEFI type.
+pub const SEPARATOR_CSTR16: &CStr16 = uefi_macros::cstr16!("\\");
+
+/// Deny list of characters for path components. UEFI supports FAT-like file
+/// systems. According to <https://en.wikipedia.org/wiki/Comparison_of_file_systems>,
+/// paths should not contain the following symbols.
+pub const CHARACTER_DENY_LIST: [Char16; 10] = unsafe {
+    [
+        Char16::from_u16_unchecked('\0' as u16),
+        Char16::from_u16_unchecked('"' as u16),
+        Char16::from_u16_unchecked('*' as u16),
+        Char16::from_u16_unchecked('/' as u16),
+        Char16::from_u16_unchecked(':' as u16),
+        Char16::from_u16_unchecked('<' as u16),
+        Char16::from_u16_unchecked('>' as u16),
+        Char16::from_u16_unchecked('?' as u16),
+        Char16::from_u16_unchecked('\\' as u16),
+        Char16::from_u16_unchecked('|' as u16),
+    ]
+};
```

Comment:
What do you think about moving the `SEPARATOR*` and `CHARACTER_DENY_LIST` constants to associated consts of `Path`? That would differ from how the [std library](https://doc.rust-lang.org/std/path/constant.MAIN_SEPARATOR.html) does it since associated constants weren't stabilized til Rust 1.20, but I think might be a slightly nicer API.

Reply:
In std::path, Path's are only a rather unvalidated frontend for paths, while the backend (fs system call API, or my file system impl in this case) actually validate this path. Because of that, I wanted to keep that separation.

What do you think?

Reply:
:+1: